### ISSUE TITLE
comp.pre futurize is broken

### ISSUE THREAD

**Issue by ericvaandering** [cms-sw/cmsdist#1849]
@ticoann @geneguvo 

I see this error with one installed from scratch too:

```
/tmp/hudson2629999002453875171.sh: /build/cmsbuild/jenkins-workarea/workspace/DMWM-WMCore-PR-27/deploy/1.0.9.pre9-comp5/sw.pre/slc6_amd64_gcc481/external/py2-future/0.15.0-comp/bin/futurize: /build/dmwmbld/srv/state/dmwmbld/builds/comp_gcc481/w/slc6_amd64_gcc481/extern: bad interpreter: No such file or directory
```

And for whatever reason the `#! python` on that script does not work.

I get around that with some gymnastics:

```
export FUTURIZE=`which futurize`
echo "#! /usr/bin/env python"|cat - $FUTURIZE > /tmp/out && mv /tmp/out $FUTURIZE
chmod +x $FUTURIZE
```

Maybe this can be folded into the RPM?

**Comment by geneguvo** [comment]
Yes. I'll fix it.

**Comment by geneguvo** [comment]
Hey Eric,
I've fixed that header line but apparently that's not enough. That script bugs trying to require 'future==0.15.0' using setuptools, which we only use to build the python packages. Even if I add it as dependency and put on environment, it is still not able to find it:

```
Traceback (most recent call last):
  File "/build/diego/w/slc6_amd64_gcc481/external/py2-future/0.15.0-comp5/bin/futurize", line 5, in <module>
    from pkg_resources import load_entry_point
  File "/build/diego/w/slc6_amd64_gcc481/external/py2-setuptools/1.1.7-comp9/lib/python2.6/site-packages/pkg_resources.py", line 2793, in <module>
    working_set.require(__requires__)
  File "/build/diego/w/slc6_amd64_gcc481/external/py2-setuptools/1.1.7-comp9/lib/python2.6/site-packages/pkg_resources.py", line 673, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/build/diego/w/slc6_amd64_gcc481/external/py2-setuptools/1.1.7-comp9/lib/python2.6/site-packages/pkg_resources.py", line 576, in resolve
    raise DistributionNotFound(req)
pkg_resources.DistributionNotFound: future==0.15.0
```

After you changed the header line, did it work for you? I've double checked my environment is all correct (I used wmcore-devtools to load it), but still the setuptools we distribute cannot load it.

**Comment by geneguvo** [comment]
I've removed the cleansing of the "egg-info" file used by setup tools, but now this futurize script fails trying to load "importlib", which is only available from py2.7:

```
$ (source w/slc6_amd64_gcc481/cms/wmcore-devtools/1.2-comp7/etc/profile.d/init.sh; futurize)
Traceback (most recent call last):
  File "/build/diego/w/slc6_amd64_gcc481/external/py2-future/0.15.0-comp2/bin/futurize", line 5, in <module>
    from pkg_resources import load_entry_point
  File "/build/diego/w/slc6_amd64_gcc481/external/py2-setuptools/1.1.7-comp9/lib/python2.6/site-packages/pkg_resources.py", line 2793, in <module>
    working_set.require(__requires__)
  File "/build/diego/w/slc6_amd64_gcc481/external/py2-setuptools/1.1.7-comp9/lib/python2.6/site-packages/pkg_resources.py", line 673, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/build/diego/w/slc6_amd64_gcc481/external/py2-setuptools/1.1.7-comp9/lib/python2.6/site-packages/pkg_resources.py", line 576, in resolve
    raise DistributionNotFound(req)
pkg_resources.DistributionNotFound: importlib
```

I'll try to add a backport of importlib to py2.6, otherwise I'll check later if it works in our py2.7 stack (not sure which python you used in your tests). Also I still did not undertand how setuptools found the future package in your setup without the egg-info file.

**Comment by geneguvo** [comment]
Ok. After adding the backported importlib library this futurize script now works on our py26 stack. I'll push the changes.

**Comment by ericvaandering** [comment]
Awesome. Yeah, I noticed the same issue with the requirements and just went back to my old way of doing it. I will give your new RPM a chance in a couple of days. (Jenkins is on 2.6 still, but Seangchan and I should try to get it running on 2.7, may be super simple).

Thanks,

Eric

**Comment by geneguvo** [comment]
Hey,
FYI. I've run it against the sitedb code base for both stage1 and stage2 changes. For stage1 it was fine, but I found some weird suggested changes as stage2. For instance, it was converting the same object to lists multiple times: ie. list(list(list(d.keys()))) instead of list(d.keys).

I've then pushed my changes to our py27 branch and re-run futurize on sitedb from there. Both stage1 and stage2 changes were fine (they matched what is documented in http://python-future.org/compatible_idioms.html).

So in summary, although the futurize script from our RPM should now work on our both py26 and py27 branches, prefer to run it from py27, at least for the stage2 changes.

**Comment by ericvaandering** [comment]
None of us are doing stage2 changes yet since that really is to get things running on python3, not just cleaning up the code to be python 2.7 compatible. It also introduces dependencies on the future module.

I think we make a push for this during LS2 which gets us on python3 before the 2020 EOL for 2.7

**Comment by ericvaandering** [comment]
The issue reported here is fixed. Closing.
